### ISSUE TITLE
Line wrapping/text reflow does not respect buffer size

### ISSUE THREAD

**Issue by ten3roberts** [pop-os/cosmic-text#219]
I have noticed that the buffer size set by `Buffer.set_size` is not always respected by the line wrapping, but overruns slightly, usually by the half end of the word.

I can reproduce this with varying texts and fonts, and discovered this when writing my GUI library which needs to limit text sizes to fit.

I have managed to find the minimal repro.

The code has the same effect if you measure the distance between the first and last glyph, so `line_w` is correct. Rendering the text also shows that it overflows despite having vertical space to wrap.

```rust
use cosmic_text::{Attrs, Buffer, FontSystem, Metrics, Shaping, Wrap};

fn main() {
    let mut font_system = FontSystem::new();
    let metrics = Metrics::new(14.0, 20.0);

    let mut buffer = Buffer::new(&mut font_system, metrics);

    let mut buffer = buffer.borrow_with(&mut font_system);

    // Add some text!
    buffer.set_wrap(Wrap::Word);
    buffer.set_text("Lorem ipsum dolor sit amet, qui minim labore adipisicing minim sint cillum sint consectetur cupidatat.", Attrs::new().family(cosmic_text::Family::Name("Inter")), Shaping::Advanced);

    // Set a size for the text buffer, in pixels
    buffer.set_size(50.0, 1000.0);

    // Perform shaping as desired
    buffer.shape_until_scroll();

    let measured_size = measure(&buffer);

    assert!(
        measured_size <= buffer.size().0,
        "Measured width is larger than buffer width\n{} <= {}",
        measured_size,
        buffer.size().0
    );
}

fn measure(buffer: &Buffer) -> f32 {
    buffer
        .layout_runs()
        .fold(0.0f32, |width, run| width.max(run.line_w))
}
```

```sh
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/cosmic-test`
thread 'main' panicked at src/main.rs:23:5:
Measured width is larger than buffer width
79.24716 <= 50
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

[Process exited 101]
```

**Comment by ten3roberts** [comment]
Example when rendering some text in a wgpu app:

![image](https://github.com/pop-os/cosmic-text/assets/25723553/32f38c35-0df9-4a76-bf56-90dcfa4f76d6)

**Comment by dtzxporter** [comment]
Did a little investigation here, It looks like this example uncovers two possible issues with layout:

```

**Comment by ten3roberts** [comment]
Awesome, I also discovered the empty run (due to unwrapping the first and last glyphs to measure width).

`fontdue` handles wrapping as best case too, and falls back to glyph breaking when there does not seem to be enough space.

However, the wrapping that I've seen overruns on are not due to insufficient space for a word, visibly, there *is* a solution where no overruns occur as there is plenty of horizontal space for multiple words, and large enough vertical space for it to wrap more lines.

**Comment by dtzxporter** [comment]
> Example when rendering some text in a wgpu app:
> 
> ![image](https://private-user-images.githubusercontent.com/25723553/300070068-32f38c35-0df9-4a76-bf56-90dcfa4f76d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY4OTkzMzYsIm5iZiI6MTcwNjg5OTAzNiwicGF0aCI6Ii8yNTcyMzU1My8zMDAwNzAwNjgtMzJmMzhjMzUtMGRmOS00YTc2LWJmNTYtOTBkY2ZhNGY3NmQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAyVDE4MzcxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliYmM4NzE3MjUxNDFiMThhODQwM2U5YmQ4Nzg1MzAxZjBmNzNhNmQ0NTNiZGE4NzE0MGNlMjBhZGNlZWUwYmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gKYrrxcChokefuDfs2DYVW9v6RZ7CypBblKUHz7auMw)

Any chance you can post the cosmic-text code that replicates this on your end? Would make it easier than fiddling with it until it breaks :D

**Comment by ten3roberts** [comment]
> > Example when rendering some text in a wgpu app:
> > 
> > ![image](https://private-user-images.githubusercontent.com/25723553/300070068-32f38c35-0df9-4a76-bf56-90dcfa4f76d6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY4OTkzMzYsIm5iZiI6MTcwNjg5OTAzNiwicGF0aCI6Ii8yNTcyMzU1My8zMDAwNzAwNjgtMzJmMzhjMzUtMGRmOS00YTc2LWJmNTYtOTBkY2ZhNGY3NmQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAyVDE4MzcxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTliYmM4NzE3MjUxNDFiMThhODQwM2U5YmQ4Nzg1MzAxZjBmNzNhNmQ0NTNiZGE4NzE0MGNlMjBhZGNlZWUwYmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gKYrrxcChokefuDfs2DYVW9v6RZ7CypBblKUHz7auMw)
> 
> Any chance you can post the cosmic-text code that replicates this on your end? Would make it easier than fiddling with it until it breaks :D

Absolutely, I am developing a UI framework https://github.com/ten3roberts/violet, which is where I discovered it.

I can push out a branch with an isolating example of a text block being wrapped. The renderer that consumes the text is rather primitive at the moment, and contained in `src/wgpu/text_renderer.rs`, along with some layout in a `TextBufferState` to share between the renderer and layout systems.

**Comment by dtzxporter** [comment]
Glyph fallback is merged into master, and I opened a PR for the empty run issue. I'll look into this overflow one as soon as I can repro it.
